File: AutomationService/AutomationService/Features/EmailVersand/Domain/Services/VersandEintrag.cs
namespace AutomationService.Features.EmailVersand.Domain.Services;

/// <summary>Auf welchem Weg die Nachricht das Haus verließ (§4.7).</summary>
public enum VersandWeg
{
    /// <summary>Die App hat über das Postfach der Kanzlei gesendet.</summary>
    Direktversand,

    /// <summary>
    /// An Outlook übergeben. <b>Nicht</b> „gesendet": Ob dort abgeschickt
    /// wurde, weiß die App nicht (§4.8) — und das Protokoll darf nichts
    /// behaupten, was es nicht weiß.
    /// </summary>
    OutlookEntwurf,

    /// <summary>Als Entwurfsdatei abgelegt, weil Outlook nicht erreichbar war.</summary>
    Entwurfsdatei,
}

/// <summary>
/// Ein Eintrag im Versandprotokoll (§4.7): wann, an wen, mit welchen Anhängen.
///
/// Für eine Kanzlei ist das der Nachweis, dass das Anspruchsschreiben hinaus
/// ist. Bis hierher hielt die App ihn nur für die Dauer eines Dialogs — danach
/// stand er nirgends mehr, und die Frage „wann ging das raus?" war nur noch im
/// Postfach zu beantworten.
/// </summary>
/// <param name="VorgangReferenz">Der Vorgang, unter dem der Eintrag hängt.</param>
/// <param name="GesendetAm">
/// Zeitpunkt der Einlieferung — beim Entwurfsweg der der Übergabe.
/// </param>
/// <param name="Weg">Direktversand, Outlook-Entwurf oder Entwurfsdatei.</param>
/// <param name="Absender">Die Adresse, von der aus gesendet wurde.</param>
/// <param name="Empfaenger">Die Adressen in „An".</param>
/// <param name="Kopie">Die Adressen in „Kopie".</param>
/// <param name="Betreff">Die Betreffzeile, wie sie hinausging.</param>
/// <param name="Anhaenge">
/// Die Namen, unter denen sie hinausgingen — nicht die auf Platte.
/// </param>
/// <param name="ImGesendetOrdner">
/// Ob die Kopie im Ordner "Gesendet" des Postfachs landete. Dort — und damit
/// in Outlook am selben Konto — liegt die Mail selbst; dieses Protokoll ist
/// der Index darüber, nicht ihr Ersatz.
/// </param>
/// <param name="MessageId">
/// Die Message-ID der eingelieferten Nachricht; null beim Entwurfsweg, wo sie
/// erst das Mailprogramm vergibt.
/// </param>
public sealed record VersandEintrag(
    string VorgangReferenz,
    DateTimeOffset GesendetAm,
    VersandWeg Weg,
    string Absender,
    IReadOnlyList<string> Empfaenger,
    IReadOnlyList<string> Kopie,
    string Betreff,
    IReadOnlyList<string> Anhaenge,
    bool ImGesendetOrdner = false,
    string? MessageId = null);

File: Automation_App_Frontend/lib/features/email_versand/domain/entities/versand_eintrag.dart
import 'package:equatable/equatable.dart';

/// Auf welchem Weg die Nachricht das Haus verließ (§4.7).
enum VersandWeg {
  /// Die App hat über das Postfach der Kanzlei gesendet.
  direktversand('Direktversand'),

  /// An Outlook übergeben — **nicht** gesendet. Ob dort abgeschickt wurde,
  /// weiß die App nicht (§4.8).
  outlookEntwurf('OutlookEntwurf'),

  /// Als Entwurfsdatei abgelegt, weil Outlook nicht erreichbar war.
  entwurfsdatei('Entwurfsdatei');

  final String kennung;

  const VersandWeg(this.kennung);

  static VersandWeg aus(String? kennung) => VersandWeg.values.firstWhere(
    (weg) => weg.kennung == kennung,
    orElse: () => VersandWeg.direktversand,
  );

  /// True nur beim Direktversand: Nur dort weiß die App, dass die Mail
  /// eingeliefert wurde. Der Unterschied gehört auf den Schirm — ein
  /// Protokoll, das eine Übergabe als Versand ausgibt, wäre als Nachweis
  /// schlechter als keines.
  bool get istNachweis => this == VersandWeg.direktversand;
}

/// Ein Versand zu einem Vorgang (§4.7): wann, an wen, mit welchen Anhängen.
///
/// Für eine Kanzlei ist das der Nachweis, dass das Anspruchsschreiben hinaus
/// ist. Die Mail selbst liegt im Ordner „Gesendet" des Postfachs — und damit
/// in Outlook am selben Konto; dieser Eintrag ist der Index darüber, nicht
/// sein Ersatz.
///
/// Das Backend liefert daneben eine `messageId`. Sie steht hier nicht: Sie
/// wäre für den Anwalt keine Auskunft, und ein Feld, das niemand liest, sieht
/// beim nächsten Blick aus wie eines, das jemand zu lesen vergessen hat.
class VersandEintrag extends Equatable {
  final String vorgangReferenz;
  final DateTime gesendetAm;
  final VersandWeg weg;
  final String absender;
  final List<String> empfaenger;
  final List<String> kopie;
  final String betreff;

  /// Die Namen, unter denen die Anhänge hinausgingen — nicht die auf Platte.
  final List<String> anhaenge;

  /// Ob die Kopie im Ordner „Gesendet" landete.
  final bool imGesendetOrdner;

  const VersandEintrag({
    required this.vorgangReferenz,
    required this.gesendetAm,
    required this.weg,
    this.absender = '',
    this.empfaenger = const [],
    this.kopie = const [],
    this.betreff = '',
    this.anhaenge = const [],
    this.imGesendetOrdner = false,
  });

  factory VersandEintrag.fromJson(Map<String, dynamic> json) => VersandEintrag(
    vorgangReferenz: json['vorgangReferenz'] as String? ?? '',
    gesendetAm:
        DateTime.tryParse(json['gesendetAm'] as String? ?? '')?.toLocal() ??
        DateTime.now(),
    weg: VersandWeg.aus(json['weg'] as String?),
    absender: json['absender'] as String? ?? '',
    empfaenger: _texte(json['empfaenger']),
    kopie: _texte(json['kopie']),
    betreff: json['betreff'] as String? ?? '',
    anhaenge: _texte(json['anhaenge']),
    imGesendetOrdner: json['imGesendetOrdner'] as bool? ?? false,
  );

  static List<String> _texte(Object? wert) => [
    for (final eintrag in (wert as List?) ?? const []) eintrag as String,
  ];

  /// Alle Angeschriebenen, „An" und „Kopie" zusammen.
  List<String> get alleEmpfaenger => [...empfaenger, ...kopie];

  @override
  List<Object?> get props => [
    vorgangReferenz,
    gesendetAm,
    weg,
    absender,
    empfaenger,
    kopie,
    betreff,
    anhaenge,
    imGesendetOrdner,
  ];
}
